File: Rotate Image.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        if n % 2 == 0:
            m = 2
        else:
            m = 3
        
        while (len(matrix) - m) >= 0:
            cur = (n - m) / 2
            for j in range(m-1):
                matrix[cur + j][-1 - cur], matrix[-1 - cur][-1 - cur - j],\
                matrix[-1 - cur - j][cur], matrix[cur][cur + j]\
                = \
                matrix[cur][cur + j], matrix[cur + j][-1 - cur],\
                matrix[-1 - cur][-1 - cur - j], matrix[-1 - cur - j][cur]
            m += 2
                
            
            

File: Rotate Image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        // for (int i = 0; i < n/2; i++)
        //     for (int j = 0 ; j < n - 1 -  2 * i; j++)
        //     {
        //         swap(matrix[i][i+j], matrix[i+j][n-1-i]);
        //         swap(matrix[i][i+j], matrix[n-1-i][n-1-i-j]);
        //         swap(matrix[i][i+j], matrix[n-1-i-j][i]);
        //     }
        for (int i = 0; i < n; i++)
            for (int j = i; j < n; j++)
                swap(matrix[i][j], matrix[j][i]);
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n/2; j++)
                swap(matrix[i][j], matrix[i][n-1-j]);
    }
};
